### PR TITLE
drivers: modem: gsm: Add optional pin control

### LINKED ISSUE #47324: drivers: modem: gsm_ppp: support common gpios
**Is your enhancement proposal related to a problem? Please describe.**
To initialize a modem successfully it is required to be in a known state.
This can be done by toggling a reset pin.

Detect when a connection is lost. This can be achieved by reading DCD.

**Describe the solution you'd like**
A solution similar to the manufacturer modem drivers specific drivers, but where none of the gpios are required.
Allow setting DTR, DCD, RI, and reset in dts.
If they are set use them in gsm_ppp.c.

**Describe alternatives you've considered**
Controlling the pins in the app layer.

### PR DESCRIPTION
**Pull request by jeppenodgaard**
Add restart (can be reset or power) and on/off timing, drt,
and wdisable pins to improve start or restart of modem.

Fixes https://github.com/zephyrproject-rtos/zephyr/issues/47324

Signed-off-by: Jeppe Odgaard <jeppe.odgaard@prevas.dk>

### CODE REVIEW

File: drivers/modem/gsm_ppp.c [hunk 1]
```diff
@@ -1124,13 +1145,62 @@ static void gsm_configure(struct k_work *work)
...
+	if (rc < 0) {
+		LOG_ERR("Failed to configure wdisable pin");
+		return rc;
+	}
+	gpio_pin_set_dt(&wdisable_gpio, 0);
+#endif
+
+#if DT_INST_NODE_HAS_PROP(0, mdm_dtr_gpios)
+	gpio_pin_set_dt(&dtr_gpio, 0);
+	rc = gpio_pin_configure_dt(&dtr_gpio, GPIO_OUTPUT_LOW);
```
Comment (gmarull) [inline drivers/modem/gsm_ppp.c:1163]: why is `_set_dt` called before configure?

File: drivers/modem/gsm_ppp.c [hunk 2]
```diff
@@ -1124,13 +1145,62 @@ static void gsm_configure(struct k_work *work)
...
+	if (rc < 0) {
+		LOG_ERR("Failed to configure dtr pin");
+		return rc;
+	}
+	gpio_pin_set_dt(&dtr_gpio, 0);
```
Comment (gmarull) [inline drivers/modem/gsm_ppp.c:1168]: why is this needed?

File: drivers/modem/gsm_ppp.c (hunk 1 above)
Comment (gmarull) [inline drivers/modem/gsm_ppp.c:1163]: you should likely use `_INACTIVE`, this is using dt flags (where you can have active low/high)

File: drivers/modem/gsm_ppp.c [hunk 3]
```diff
@@ -121,6 +121,27 @@ NET_BUF_POOL_DEFINE(gsm_recv_pool, GSM_RECV_MAX_BUF, GSM_RECV_BUF_SIZE, 0, NULL)
...
+#endif /* DT_INST_NODE_HAS_PROP(0, mdm-restart-gpios) */
+
+#if DT_INST_NODE_HAS_PROP(0, mdm_dtr_gpios)
+static const struct gpio_dt_spec dtr_gpio =
+			GPIO_DT_SPEC_INST_GET(0, mdm_dtr_gpios);
+#endif
+#if DT_INST_NODE_HAS_PROP(0, mdm_wdisable_gpios)
+static const struct gpio_dt_spec wdisable_gpio =
+			GPIO_DT_SPEC_INST_GET(0, mdm_wdisable_gpios);
+#endif
```
Comment (gmarull) [inline drivers/modem/gsm_ppp.c:143]: please use driver's config instead (to align with how drivers are implemented in general)

File: drivers/modem/gsm_ppp.c [hunk 4]
```diff
@@ -1124,13 +1145,62 @@ static void gsm_configure(struct k_work *work)
...
+	}
+	LOG_INF("Stop modem");
+	gpio_pin_set_dt(&restart_gpio, 1);
+	k_sleep(K_MSEC(DT_INST_PROP(0, mdm_off_delay)));
+
+	LOG_INF("Start modem");
+	gpio_pin_set_dt(&restart_gpio, 0);
+	k_sleep(K_MSEC(DT_INST_PROP(0, mdm_on_delay)));
+
+	LOG_INF("... Done!");
```
Comment (gmarull) [inline drivers/modem/gsm_ppp.c:1185]: `LOG_DBG`?

**Review by gmarull — CHANGES_REQUESTED** [review]

**Comment by ycsin** [comment]
There was a consensus last time to keep these power on/off sequence out of the GSM ppp modem as this driver is shared by a lot of different manufacturers and each of them have slightly different on/off sequence. An agreed & better way would be to create APIs to register such on/off functions to be called when the modem entered certain states, I remember such PRs should be available, probably closed due to inactivity

**Comment by jeppenodgaard** [comment]
@ycsin thanks for enlightning me. I was not aware of those PRs.

I think these are the PRs you recall. The first PR has a good debate and leads to the second PR:

- https://github.com/zephyrproject-rtos/zephyr/pull/38317.
- https://github.com/zephyrproject-rtos/zephyr/pull/38946.

I though it was possible to add something generic to the dts, but it seems like it is not the case because a lot of modems has something unique to them. So toggling a power or reset pin with custom delays unfortunately will not always suffice.

I did not notice the need for implementing modem_on_cb and modem_off_cb. This is what I need to do.
It was not clear to me that these were callbacks for the application. Maybe it can be made clearer?

Also is there a simple way to define gpios in the dts which is needed by the application or do I need to add a out-of-tree driver?

**Comment by ycsin** [comment]
The GSM modem sample should already make it quite clear that it is a power on/off cb, the only improvements that I think would be helpful is to have the cb to control the digital pins. Afaik the closest thing that you can refer to is the blinky sample.

**Comment by jeppenodgaard** [comment]
True. It is clear that those are ment for the application.

What I am trying to get at is that maybe it would make sense if I made a PR and tried to document what the callbacks are ment to be used for and maybe with some use cases listed..?

**Comment by ycsin** [comment]
PRs are always welcomed, everyone here appreciates clearer documentation and moar samples.
